azure/datalake/store/transfer.py

```diff
@@ -240,9 +245,14 @@ def __init__(self, adlfs, transfer, merge=None, nthreads=None,
         self._chunked = chunked
         self._unique_temporary = unique_temporary
         self._unique_str = uuid.uuid4().hex
+        self._progress_callback=progress_callback
+        self._progress_lock = threading.Lock()
```

Comment:
@brettcannon thank you for the hints and details about the memory model. Being a casual Python user, this is helpful.

I thought about something similar as an alternative (I was thinking about keeping the progress as part of the existing chunk state. I'd be ok with either way, if this proves to have an impact on performance. If it doesn't I'd lean towards the simpler implementation, but glad to do whatever you think it's best.

Not quite familiar with the Python locking implementation but was assuming a lightweight approach initially and an uncontended lock may be cheap in this case. Even with a large number of small files, we're still limited by the actual number of threads that are performing IO operations in the order of milliseconds at least while an unconteded lock seems to be at least two orders of magnitude faster (0.1 us - https://mail.python.org/pipermail/python-ideas/2010-July/007644.html).

 I'd run some tests just to satisfy my curiosity.